text/0000-type-declarations.md

```diff
@@ -0,0 +1,128 @@
+- Start Date: 2018-02-21
+- RFC PR:
+- WordPress Coding Standards Issue:
+
+# Summary
+
+[Type Declarations](http://php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration) make code more readable. They also protect code, because Type Declarations are a way of enforcing specific data types in parameters passed to a function.
+
+# Basic Example
+
+You can force callers to pass an `array`, or a specific type of object, such as a `WP_REST_Request`. If a caller doesn't pass the data type expected, PHP 5 triggers a recoverable fatal error and PHP 7 throws a [TypeError](http://php.net/manual/en/class.typeerror.php) exception. This behavior catches problems right away.
+
+```php
+protected function foo( Bar $bar, array $items ) {
+  // $bar   must be an instance of the Bar class.
+  // $items must be passed as an array.
+}
+```
+
+# Motivation
+
+PHP is a [weakly typed language](https://en.wikipedia.org/wiki/Strong_and_weak_typing). It doesn't require you to declare data types. However, variables still have data types associated with them (e.g., `string`, `integer`). In a weakly typed language you can do radical things like adding a `string` to an `integer` with no error. While this can be wonderful in some cases, it can also lead to unanticipated behavior and bugs.
```

Comment:
>... In a weakly typed language, you can do radical things like adding a `string` to an `integer` with no error.

Is this really radical? :D Maybe this should be reworded. Type juggling is a known thing in weakly typed languages

Reply:
Good point. I agree that this word is a little too strong.